### PR TITLE
Feature/ops wcoss2

### PR DESCRIPTION
**Pull request by BoiVuong-NOAA**
Issues #399  
GFS in the branch feature feature/ops-wcoss2 transition to WCOSS2

- 1. gfs drivers - all drivers have modified to run on Cactus (WCOSS2)
- 2. j-jobs - all j-jobs have modified COMIN and COMOUT and replaced mpirun to mpiexec
- 3. ex-scripts - exgfs_atmos_gempak_meta.sh has modified to mpirun to mpiexec

### CODE REVIEW

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+# Now set up GEMPAK/NTRANS environment
+###########################################
+module load gempak/$gempak_ver
+module list
+
+############################################
+# Define COM, COMOUTwmo, COMIN  directories
+############################################
+# set envir=para or para to test with data in prod or para
+ export envir=para
```
Comment (KateFriedman-NOAA) [inline driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:70]: Remove duplicate line.

Comment (BoiVuong-NOAA) [reply driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:70]: This setting in drivers are for my testing.  You can remove it or leave it as is.    The script will pick it up as second of variable "envir".    The purpose of this variable is to allow me to use the data from PARA or PROD for testing.

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+#############################################################
+# Specify Run Driver location
+#############################################################
+export driver=/lfs/h2/emc/vpppg/noscrub/$USER/packages/gfs.$gfs_ver/driver
+
+#############################################################
+# Load module version on WCOSS2
+#############################################################
+. ${driver}/product/run.version
```
Comment (KateFriedman-NOAA) [inline driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:42]: A new run.ver file should exist in a new versions subfolder at the top level (`versions/run.ver`). Please update this line for that. Please also rename the run.version file you created to `run.ver` and move it up and into a new subfolder at the top level called `versions`. I'll be placing a new build.ver file in that new subfolder as well. See slide 9 of this slidedeck for what I'm referring to regarding the versions subfolder:
https://docs.google.com/presentation/d/15v-7rEM2CkJlEzwX4sE_qJd8DFJCgI726pzghIM_AwE/edit#slide=id.ge5cc65791d_0_31
Thanks!

File: driver/product/run.version [hunk 3]
```diff
@@ -0,0 +1,37 @@
+export version=v16.0.0
+export gfs_ver=v16.0.0
```
Comment (KateFriedman-NOAA) [inline driver/product/run.version:2]: The GFS will be `v16.2.0` once in ops on WCOSS2.

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+export MP_LABELIO=yes
+
+set -xa
+export PS4='$SECONDS + '
+date
+
+#############################################################
+# Specify GFS versions
+#############################################################
+export gfs_ver=v16.0.0
```
Comment (KateFriedman-NOAA) [inline driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:32]: The GFS will be `v16.2.0` once in ops on WCOSS2.

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
Updates look good, left a couple comments regarding run.ver. Will approve and submit once comments are addressed. Thanks!

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 (hunk 1 above)
Comment (KateFriedman-NOAA) [reply driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:70]: @BoiVuong-NOAA It's fine to have envir=para in the scripts for your use...I'd just like the duplicate one on line 70 removed for cleanliness. Please check the other scripts and remove similar duplicate lines. Thanks!

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 (hunk 4 above)
Comment (BoiVuong-NOAA) [reply driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:32]: Thanks, Kate. I will update once GFS GFS will be v16.2.0 once in ops on WCOSS2.

**Comment by KateFriedman-NOAA** [comment]
@BoiVuong-NOAA Please update your fork branch copy of feature/ops-wcoss2, I submitted two PRs into it with changes from my own fork branch. Click the "Fetch Upstream" button in GitHub when viewing your fork branch. That will sync your branch with the auth copy ahead of me submitting this PR. Thanks!

**Comment by BoiVuong-NOAA** [comment]
Thanks, Kate.
I have updated my fork branch copy of feature/ops-wcoss2.

File: driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00 (hunk 2 above)
Comment (KateFriedman-NOAA) [reply driver/product/run_JGDAS_ATMOS_GEMPAK_META_NCDC_wcoss2.sh_00:42]: @BoiVuong-NOAA I am going to submit this PR and then move run.version as I mentioned in my prior comment. Thanks!

**Review by KateFriedman-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
@BoiVuong-NOAA @GuangPingLou-NOAA Please sync your fork branches now that this PR has gone in, thanks!

**Comment by BoiVuong-NOAA** [comment]
Thanks, Kate.  I will delete this fork of NOAA-EMC/global-workflow.

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA I fetched the Authoritative branch ops-wcoss2. I didn't see file "wcoss2.env".

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA Okay, Now I see the file "wcoss2.env".

**Comment by GuangPingLou-NOAA** [comment]
@KateFriedman-NOAA But in the file, "export launcher="mpiexec -l", I thought it should be export launcher="mpiexec -n". In my case, I need to have "mpiexec -n $npe_postsnd --depth=6 --cpu-bind depth", with PBS assignemnt:
"#PBS -l place=vscatter:excl,select=2:ncpus=128:mpiprocs=20:mem=500GB". 
It needs 40 tasks in 2 nodes with 128 tasks per node, exclusive, Memory is 1,000GB in total.

**Comment by GuangPingLou-NOAA** [comment]
where npe_postsnd=40.

**Comment by KateFriedman-NOAA** [comment]
@GuangPingLou-NOAA If you look at how the $launcher variable is used in WCOSS2.env you'll see that the `-n` is now set in each job-specific part of the if-block below. On WCOSS2 we need to specify `-n` and `-ppn` so I moved the `-n` down into the job-specific sections and added the requested `-l` to the launcher setting.. The bulk of the APRUN* variables will be set like this now:

```
${launcher} -n $npe_JOB -ppn $npe_node_JOB ...
```

...which will look like this at runtime:

```
mpiexec -l -n ### -ppn ### ...
```

You said for the postsnd job you need:
```
mpiexec -n $npe_postsnd --depth=6 --cpu-bind depth
```

To get that from the [WCOSS2.env postsnd section](https://github.com/GuangPingLou-NOAA/global-workflow/blob/feature/ops-wcoss2/env/WCOSS2.env#L209) you can try setting this in the env file:
```
export APRUN_POSTSND="$launcher -n $npe_postsnd -ppn $npe_node_postsnd --depth=6 --cpu-bind depth"
```

^ that will look like this at runtime: `mpiexec -l -n ### -ppn ### --depth=6 --cpu-bind depth`

Also update both parm/config/config.resource* configs to set the resources for the postsnd. Here is the section to adjust in config.resources.emc.dyn:

https://github.com/GuangPingLou-NOAA/global-workflow/blob/feature/ops-wcoss2/parm/config/config.resources.emc.dyn#L391:L404

Let me know if you have any questions...post them in issue #399 or through Slack or through email. This PR is closed now. Thanks!

**Comment by GuangPingLou-NOAA** [comment]
Thanks, Kate.
I'll test it out.

**Pull request merged**